sharktank/sharktank/ops/sharded_impls.py

```diff
@@ -1126,19 +1126,30 @@ def mean_split(
         ]
         return SplitPrimitiveTensor(ts=shards, shard_dim=shard_dim_new)
     else:
-        gathered = cat(
-            [
-                (
-                    transfer_to_logical_device(shard, x.devices[0])
-                    if i != 0
-                    else barrier_on_logical_device(shard, x.devices[0])
-                )
-                for i, shard in enumerate(x.shards)
-            ],
-            dim=x.shard_dim,
+        partial_sums = [
+            sum(shard, dim=dim, keepdim=keepdim, dtype=dtype) for shard in x.shards
+        ]
+
+        partial_cnts = [
+            torch.tensor(
+                math.prod(shard.shape[d] for d in dim),
+                device=unbox_tensor(shard).device,
+                dtype=dtype,
+            )
+            for shard in x.shards
+        ]
+        sums_spt = SplitPrimitiveTensor(ts=partial_sums, shard_dim=0, devices=x.devices)
+        cnts_on_hub = [barrier_on_logical_device(partial_cnts[0], x.devices[0])] + [
+            transfer_to_logical_device(ct, x.devices[0]) for ct in partial_cnts[1:]
+        ]
+
+        global_sum = sharded_sum(sums_spt)
+        global_cnt = torch.stack(cnts_on_hub).sum()
```

Comment:
I would manually iterate and sum rather than stack and sum. It is counter intuitive but stacking then summing will likely be slower.